privacyidea/lib/periodictask.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+#  2018-06-25 Friedrich Weber <friedrich.weber@netknights.it>
+#             Initial implementation of periodic tasks
+#
+# This code is free software; you can redistribute it and/or
+# modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+__doc__ = """This module provides functions to manage periodic tasks in the database,
+to determine their next scheduled running time and to run them."""
+
+import logging
+from datetime import datetime
+
+from croniter import croniter
+from dateutil.tz import tzutc, tzlocal
+
+from privacyidea.lib.error import ServerError, ParameterError
+from privacyidea.models import PeriodicTask
+
+log = logging.getLogger(__name__)
+
+
+def calculate_next_timestamp(ptask, node, interval_tzinfo=None):
+    """
+    Calculate the timestamp of the next scheduled run of task ``ptask`` on node ``node``.
+    We do not check if the task is even scheduled to run on the specified node.
+    If the periodic task has no prior run recorded on the specified node, a
+    ``ServerError`` is thrown. Malformed cron expressions may throw a
+    ``ValueError``.
+
+    :param ptask: Dictionary describing the periodic task, as from ``PeriodicTask.get()``
+    :param node: Node on which the periodic task is scheduled
+    :type node: unicode
+    :param interval_tzinfo: Timezone in which the cron expression should be interpreted. Defaults to local time.
+    :type interval_tzinfo: tzinfo
+    :return: a timezone-aware (UTC) datetime object
+    """
+    if interval_tzinfo is None:
+        interval_tzinfo = tzlocal()
+    if node not in ptask["last_runs"]:
+        raise ServerError("No last run on node {!r} recorded for task {!r}".format(node, ptask["name"]))
```

Comment:
At one point a task has to run for the first time.
could this cause problems to throw an exception, if the task did not run before (first-runner)

Reply:
You're right, I forgot to mention this: In order to calculate the timestamp of the next run, we always need the last run as a reference point.
My current approach (I have implemented this locally already) is to add a ``last_run`` when a new periodic task is created whose time is the time of the task creation. This is not so nice, because the task has not actually run at that time -- but calculating the next scheduled run works fine then.